snapcraft/internal/elf.py

```diff
@@ -35,6 +35,72 @@
 logger = logging.getLogger(__name__)
 
 
+class ElfFile:
+    """ElfFile represents and elf file on a path and its attributes."""
+
+    def __init__(self, *, path: str, is_executable: bool) -> None:
+        """Initialize an ElfFile instance.
+
+        :param str path: path to an elf_file within a snapcraft project.
+        :param bool is_executable: True if the elf_file is an executable,
+                                   meaning that it has in .interp entry
+                                   in its headers.
+        """
+        self.path = path
+        self.is_executable = is_executable
+
+
+class Patcher:
+    """Patcher holds the necessary logic to patch elf files."""
+
+    def __init__(self, *, dynamic_linker: str) -> None:
+        """Create a Patcher instance.
+
+        :param str dynamic_linker: the path to the dynamic linker to set the
+                                   elf file to.
+        """
+        self._dynamic_linker = dynamic_linker
+
+        # If we are running from the snap we want to use the patchelf
+        # bundled there as it would have the capabilty of working
+        # anywhere given the fixed ld it would have.
+        # If not found, resort to whatever is on the system brought
+        # in by packaging dependencies.
+        if common.is_snap():
+            snap_dir = os.getenv('SNAP')
+            self._patchelf_cmd = os.path.join(snap_dir, 'bin', 'patchelf')
+        else:
+            self._patchelf_cmd = 'patchelf'
+
+    def patch(self, *, elf_file: ElfFile) -> None:
+        """Patch elf_file with the Patcher instance configuration.
+
+        patch will to the right thing with regards to the different types
```

Comment:
s/to/do/, although I must say this isn't the most helpful comment :stuck_out_tongue: . Perhaps something like:

"If the ELF is executable, patch it to use the configured linker."